File: near/src/main/java/com/adroitandroid/near/connect/NearConnect.java
package com.adroitandroid.near.connect;

import android.content.Context;
import android.os.Looper;

import com.adroitandroid.near.discovery.NearDiscovery;
import com.adroitandroid.near.model.Host;

import java.util.Set;

/**
 * Created by pv on 21/06/17.
 */

public interface NearConnect {

    long send(byte[] bytes, Host peer);

    void startReceiving();

    void stopReceiving(boolean abortCurrentTransfers);

    Set<Host> getPeers();

    boolean isReceiving();

    class Builder {
        private Context mContext;
        private Listener mListener;
        private Looper mListenerLooper;
        private Set<Host> mPeers;

        public Builder setContext(Context context) {
            mContext = context;
            return this;
        }

        public Builder setListener(Listener listener, Looper listenerLooper) {
            mListener = listener;
            mListenerLooper = listenerLooper;
            return this;
        }

        public Builder fromDiscovery(NearDiscovery discovery) {
            if (discovery != null) {
                mPeers = discovery.getAllAvailablePeers();
            } else {
                mPeers = null;
            }
            return this;
        }

        public Builder forPeers(Set<Host> peers) {
            mPeers = peers;
            return this;
        }

        public NearConnect build() {
            return new NearConnectImpl(mContext, mListener, mListenerLooper, mPeers);
        }
    }

    interface Listener {
        void onReceive(byte[] bytes, Host sender);

        void onSendComplete(long jobId);

        void onSendFailure(Throwable e, long jobId);

        void onStartListenFailure(Throwable e);
    }
}

File: near/src/main/java/com/adroitandroid/near/connect/NearConnect.kt
package com.adroitandroid.near.connect

import android.content.Context
import android.os.Looper
import com.adroitandroid.near.connect.server.TcpServerService
import com.adroitandroid.near.discovery.NearDiscovery
import com.adroitandroid.near.model.Host

interface NearConnect {
    fun send(bytes: ByteArray, peer: Host): Long
    fun startReceiving()
    fun stopReceiving(abortCurrentTransfers: Boolean)
    val peers: Set<Host>
    val isReceiving: Boolean

    class Builder {
        private lateinit var mContext: Context
        private lateinit var mListener: Listener
        private lateinit var mListenerLooper: Looper
        private lateinit var mPeers: Set<Host>
        private var mPort: Int = TcpServerService.SERVER_PORT

        fun setContext(context: Context): Builder {
            mContext = context
            return this
        }

        fun setListener(listener: Listener, listenerLooper: Looper): Builder {
            mListener = listener
            mListenerLooper = listenerLooper
            return this
        }

        fun fromDiscovery(discovery: NearDiscovery): Builder {
            mPeers = discovery.allAvailablePeers
            return this
        }

        fun forPeers(peers: Set<Host>): Builder {
            mPeers = peers
            return this
        }

        fun setPort(port: Int): Builder {
            mPort = port
            return this
        }

        fun build(): NearConnect {
            return NearConnectImpl(mContext, mListener, mListenerLooper, mPeers, mPort)
        }
    }

    interface Listener {
        fun onReceive(bytes: ByteArray, sender: Host)
        fun onSendComplete(jobId: Long)
        fun onSendFailure(e: Throwable?, jobId: Long)
        fun onStartListenFailure(e: Throwable?)
    }
}